File: backend/app/ai_interview/utils/timecode.py
"""Timecode utilities for video/audio processing"""
from typing import Union


class Timecode:
    """Timecode utilities for converting between formats"""
    
    @staticmethod
    def seconds_to_hms(seconds: float) -> str:
        """Convert seconds to HH:MM:SS.mmm format"""
        hours = int(seconds // 3600)
        minutes = int((seconds % 3600) // 60)
        secs = int(seconds % 60)
        millis = int((seconds % 1) * 1000)
        return f"{hours:02d}:{minutes:02d}:{secs:02d}.{millis:03d}"
    
    @staticmethod
    def hms_to_seconds(hms: str) -> float:
        """Convert HH:MM:SS.mmm format to seconds"""
        parts = hms.split(":")
        if len(parts) != 3:
            raise ValueError("Invalid timecode format")
        
        hours = int(parts[0])
        minutes = int(parts[1])
        sec_parts = parts[2].split(".")
        seconds = int(sec_parts[0])
        millis = int(sec_parts[1]) if len(sec_parts) > 1 else 0
        
        return hours * 3600 + minutes * 60 + seconds + millis / 1000.0
    
    @staticmethod
    def format_for_ffmpeg(seconds: float) -> str:
        """Format seconds for ffmpeg duration/seek parameters"""
        return Timecode.seconds_to_hms(seconds)
    
    @staticmethod
    def clamp(value: float, min_val: float, max_val: float) -> float:
        """Clamp value between min and max"""
        return max(min_val, min(max_val, value))

File: frontend/src/lib/timecode.ts
/**
 * Timecode utilities for video/audio processing
 */

export function formatTimecode(seconds: number | string): string {
  // Ensure we have a number
  const numSeconds = typeof seconds === 'string' ? Number(seconds) : seconds;
  if (isNaN(numSeconds) || !isFinite(numSeconds)) {
    return '0:00.000';
  }
  
  const hours = Math.floor(numSeconds / 3600);
  const minutes = Math.floor((numSeconds % 3600) / 60);
  const secs = Math.floor(numSeconds % 60);
  const millis = Math.floor((numSeconds % 1) * 1000);
  
  if (hours > 0) {
    return `${hours}:${minutes.toString().padStart(2, '0')}:${secs.toString().padStart(2, '0')}.${millis.toString().padStart(3, '0')}`;
  }
  return `${minutes}:${secs.toString().padStart(2, '0')}.${millis.toString().padStart(3, '0')}`;
}

export function parseTimecode(timecode: string): number {
  const parts = timecode.split(':');
  if (parts.length === 3) {
    const hours = parseInt(parts[0], 10);
    const minutes = parseInt(parts[1], 10);
    const secParts = parts[2].split('.');
    const seconds = parseInt(secParts[0], 10);
    const millis = parseInt(secParts[1] || '0', 10);
    return hours * 3600 + minutes * 60 + seconds + millis / 1000;
  } else if (parts.length === 2) {
    const minutes = parseInt(parts[0], 10);
    const secParts = parts[1].split('.');
    const seconds = parseInt(secParts[0], 10);
    const millis = parseInt(secParts[1] || '0', 10);
    return minutes * 60 + seconds + millis / 1000;
  }
  return 0;
}

export function clamp(value: number, min: number, max: number): number {
  return Math.max(min, Math.min(max, value));
}

